### ISSUE TITLE
BaseIntegratorMove does not distinguish between reasons for Exceptions

### ISSUE THREAD

**Issue by jchodera** [choderalab/openmmtools#643]
Currently, [`BaseIntegratorMove`](https://github.com/choderalab/openmmtools/blob/2e64980d86cf087f69e282927eebf2b41b73a3ea/openmmtools/mcmc.py#L601), which is the base class of classes like [`LangevinDynamicsMove`](https://github.com/choderalab/openmmtools/blob/2e64980d86cf087f69e282927eebf2b41b73a3ea/openmmtools/mcmc.py#L1018), [does not distinguish between causes for `Exceptions` during dynamics](https://github.com/choderalab/openmmtools/blob/2e64980d86cf087f69e282927eebf2b41b73a3ea/openmmtools/mcmc.py#L718). Meant to guard against NaNs, this behavior means that even important failures that should be relayed to the user, like
```
openmm.OpenMMException: The periodic box size has decreased to less than twice the nonbonded cutoff.
```
instead trigger restarts rather than useful warnings. We should instead catch behavior like this that is important to communicate to users because it can frustrate debugging.

**Comment by ijpulidos** [comment]
We should just try with one of the testsystems and set the box size to less than twice the nonbonded cutoff and reproduce the behavior.

**Comment by ijpulidos** [comment]
I am struggling to write a good example that fails in this step. Since if I define some test system to have a box size smaller than twice of the cutoff distance it fails in the context creation in a different part of the code. My guess here is that we actually need some system that doesn't raise an exception on setup but NaNs along the way.

**Comment by Dan-Burns** [comment]
I attached a system to issue [484](https://github.com/choderalab/openmmtools/issues/484) that might be useful for identifying NaNs.

I haven't seen an issue posted that's similar to mine and I've run into it so many times that I feel like I have to be missing something obvious.

Any replica exchange system I try to run errors out with one of the replicas (usually corresponding to a thermodynamic state with a lambda above 0.9) despite providing a well equilibrated starting structure:
`Potential energy is NaN after 4 attempts of integration with move LangevinSplittingDynamicsMove
Propagating replica 3 at state 3 resulted in a NaN!
`

I thought I resolved it by minimizing the replica exchange ensemble and doing a brief equilibration with a really high collision rate and really small timestep and then feeding that simulation.sampler_states to a new ReplicaExchangeSampler but I have just run into the issue again even with this approach.  I don't know if it could be hardware dependent, but I can run the same script on a local gpu and avoid the error (although that might have been a lucky random seed?).

Thank you,

Dan
